# Lecture Note: Week 06 — Function (ระดับพื้นฐาน)

เอกสารฉบับนี้เป็น **เอกสารประกอบการสอน (lecture note)** แยกจาก slide deck `slides/06_function.md`
มีจุดประสงค์เพื่อขยายความแนวคิดเรื่องฟังก์ชัน (function) ให้ละเอียดกว่าที่นำเสนอในสไลด์ 7 หน้า
โดยเนื้อหาในไฟล์นี้ครอบคลุม **5 หัวข้อระดับพื้นฐาน** ซึ่งเป็น **แกนเนื้อหาที่ตรงกับสิ่งที่สอนจริงในชั้นเรียน
และอยู่ในขอบเขตของข้อสอบวิชา Programming 1 สัปดาห์ที่ 6 ทั้งหมด**

เนื้อหาทั้ง 5 หัวข้อในไฟล์นี้ ได้แก่

1. จุดประสงค์ของฟังก์ชัน (DRY, decomposition)
2. Function definition กับ Function call
3. Parameter กับ Argument
4. Return value กับ Procedure (void function)
5. ขอบเขตตัวแปร (scope): ตัวแปรเฉพาะที่ (local) กับตัวแปรส่วนกลาง (global)

แนวคิดทุกหัวข้อจะอธิบายแบบ **language-agnostic** (ไม่ผูกกับภาษาใดภาษาหนึ่ง) ก่อน แล้วจึงแสดงตัวอย่างด้วยภาษา Python
เนื่องจากเป็นภาษาที่ใช้สอนในคอร์สนี้ ตัวอย่างโค้ดทุกตัวอย่างผ่านการรันจริงด้วย Python 3.11.15 แล้ว
(ดูรายละเอียดวิธีการยืนยันผลลัพธ์ในหมายเหตุท้ายไฟล์)

ศัพท์เทคนิคในเอกสารนี้ยึดตามตารางศัพท์มาตรฐานที่กำหนดไว้ใน `plan.md` เช่น function = ฟังก์ชัน,
parameter = พารามิเตอร์, argument = อาร์กิวเมนต์, scope = ขอบเขต (ตัวแปร), local variable = ตัวแปรเฉพาะที่ (local),
global variable = ตัวแปรส่วนกลาง (global)

---

## 1: จุดประสงค์ของฟังก์ชัน (DRY, Decomposition)

### แนวคิด

ฟังก์ชัน (function) คือกลุ่มคำสั่งที่ถูกตั้งชื่อไว้และสามารถเรียกใช้ซ้ำได้หลายครั้งโดยไม่ต้องเขียนคำสั่งชุดเดิมซ้ำใหม่ทุกครั้ง
จุดประสงค์หลักของการใช้ฟังก์ชันมีสองประการที่สัมพันธ์กัน ประการแรกคือหลักการ **DRY (Don't Repeat Yourself)**
ซึ่งหมายถึงการเขียนตรรกะการทำงานหนึ่งอย่างไว้เพียงจุดเดียวในโปรแกรม แล้วเรียกใช้ซ้ำเมื่อจำเป็น
แทนที่จะคัดลอกวางคำสั่งชุดเดิมไปไว้หลายจุด เพราะหากต้องแก้ไขตรรกะนั้นในภายหลัง การมีอยู่จุดเดียวจะทำให้แก้ไขได้ง่ายและลดโอกาสเกิดข้อผิดพลาด
ประการที่สองคือ **decomposition** หรือการแตกปัญหาใหญ่ที่ซับซ้อนให้กลายเป็นปัญหาย่อยหลายส่วน โดยแต่ละส่วนเขียนเป็นฟังก์ชันของตนเอง
ทำให้โปรแกรมอ่านง่ายขึ้น ทดสอบแยกส่วนได้ และดูแลรักษาง่ายขึ้นในระยะยาว

ในภาษา Python ฟังก์ชันถูกสร้างขึ้นด้วยคำสั่ง `def` ตามด้วยชื่อฟังก์ชันและวงเล็บ ดังจะเห็นในตัวอย่างถัดไป
ซึ่งแสดงให้เห็นว่าเมื่อมีฟังก์ชันแล้ว การนำตรรกะเดิมไปใช้ซ้ำทำได้เพียงเรียกชื่อฟังก์ชัน ไม่ต้องเขียนสูตรคำนวณซ้ำใหม่ทุกครั้ง

### ตัวอย่าง Python

#### ตัวอย่างที่ 1.1 — แนวคิดหลัก: ใช้ฟังก์ชันเดียวคำนวณซ้ำหลายครั้ง (DRY)

In [1]:
def calculate_area(width, height):
    return width * height

print(calculate_area(4, 5))
print(calculate_area(6, 2))
print(calculate_area(3, 3))

20
12
9


**Output:**
```
20
12
9
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 4 | เรียก `calculate_area(4, 5)` | width=4, height=5 | เข้าสู่ฟังก์ชัน ครั้งที่ 1 |
| 2 | `return width * height` | คืนค่า 20 | คำนวณพื้นที่และส่งกลับ |
| 4 | `print(...)` | แสดงผล 20 | พิมพ์ผลลัพธ์ครั้งที่ 1 |
| 5 | เรียก `calculate_area(6, 2)` | width=6, height=2 | เข้าสู่ฟังก์ชัน ครั้งที่ 2 ใช้โค้ดเดิมโดยไม่ต้องเขียนใหม่ |
| 2 | `return width * height` | คืนค่า 12 | คำนวณพื้นที่และส่งกลับ |
| 5 | `print(...)` | แสดงผล 12 | พิมพ์ผลลัพธ์ครั้งที่ 2 |
| 6 | เรียก `calculate_area(3, 3)` | width=3, height=3 | เข้าสู่ฟังก์ชัน ครั้งที่ 3 |
| 2 | `return width * height` | คืนค่า 9 | คำนวณพื้นที่และส่งกลับ |
| 6 | `print(...)` | แสดงผล 9 | พิมพ์ผลลัพธ์ครั้งที่ 3 |

**คำอธิบายภาพรวม:** บรรทัด 1–2 นิยามฟังก์ชัน `calculate_area` เพียงครั้งเดียว บรรทัด 4–6 เรียกใช้ฟังก์ชันเดิมซ้ำ 3 ครั้งด้วยอาร์กิวเมนต์ต่างกัน
แสดงหลักการ DRY อย่างชัดเจน คือเขียนตรรกะการคำนวณพื้นที่ (`width * height`) ไว้ที่เดียว แล้วนำกลับมาใช้ซ้ำได้เรื่อย ๆ
แทนที่จะคัดลอกสูตรไปเขียนซ้ำสามแห่ง ซึ่งหากภายหลังต้องแก้สูตรคำนวณ ก็แก้เพียงจุดเดียวคือบรรทัด 2

#### ตัวอย่างที่ 1.2 — บริบทอื่นที่ต่างโดเมน: DRY กับงานจัดรูปแบบข้อความ (ไม่ใช่แค่คำนวณเลข)

In [2]:
def format_receipt_line(item, price):
    return f"{item} ราคา {price} บาท"

print(format_receipt_line("กาแฟ", 45))
print(format_receipt_line("ขนมปัง", 25))
print(format_receipt_line("น้ำส้ม", 30))

กาแฟ ราคา 45 บาท
ขนมปัง ราคา 25 บาท
น้ำส้ม ราคา 30 บาท


**Output:**
```
กาแฟ ราคา 45 บาท
ขนมปัง ราคา 25 บาท
น้ำส้ม ราคา 30 บาท
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 4 | เรียก `format_receipt_line("กาแฟ", 45)` | item="กาแฟ", price=45 | เข้าสู่ฟังก์ชัน ครั้งที่ 1 |
| 2 | `return f"..."` | คืนค่า "กาแฟ ราคา 45 บาท" | ประกอบข้อความตามรูปแบบเดียวกันทุกครั้ง |
| 4 | `print(...)` | แสดงผล กาแฟ ราคา 45 บาท | พิมพ์ผลลัพธ์ครั้งที่ 1 |
| 5 | เรียก `format_receipt_line("ขนมปัง", 25)` | item="ขนมปัง", price=25 | เข้าสู่ฟังก์ชัน ครั้งที่ 2 ใช้โค้ดเดิมโดยไม่ต้องเขียนใหม่ |
| 2 | `return f"..."` | คืนค่า "ขนมปัง ราคา 25 บาท" | ประกอบข้อความตามรูปแบบเดียวกัน |
| 5 | `print(...)` | แสดงผล ขนมปัง ราคา 25 บาท | พิมพ์ผลลัพธ์ครั้งที่ 2 |
| 6 | เรียก `format_receipt_line("น้ำส้ม", 30)` | item="น้ำส้ม", price=30 | เข้าสู่ฟังก์ชัน ครั้งที่ 3 |
| 2 | `return f"..."` | คืนค่า "น้ำส้ม ราคา 30 บาท" | ประกอบข้อความตามรูปแบบเดียวกัน |
| 6 | `print(...)` | แสดงผล น้ำส้ม ราคา 30 บาท | พิมพ์ผลลัพธ์ครั้งที่ 3 |

**คำอธิบายภาพรวม:** บรรทัด 1–2 นิยามฟังก์ชัน `format_receipt_line` ที่รับพารามิเตอร์ `item` และ `price` แล้วประกอบเป็นข้อความรายการใบเสร็จรูปแบบเดียวกันเสมอ
บรรทัด 4–6 เรียกใช้ฟังก์ชันเดิมซ้ำ 3 ครั้งด้วยรายการสินค้าและราคาต่างกัน แสดงให้เห็นว่าหลักการ DRY ที่สาธิตในตัวอย่างที่ 1.1
(ซึ่งเป็นโดเมนตัวเลขเชิงเรขาคณิต) ใช้ได้ผลเช่นเดียวกันกับงานด้านข้อความ (text) เช่นการจัดรูปแบบรายการใบเสร็จ ไม่ได้จำกัดอยู่แค่การคำนวณตัวเลขเท่านั้น

#### ตัวอย่างที่ 1.3 — ข้อผิดพลาดที่พบบ่อย: ไม่ทำ DRY แล้วคัดลอกสูตรผิด

In [3]:
def celsius_to_fahrenheit_v1(celsius):
    return celsius * 9 / 5 + 32

def celsius_to_fahrenheit_v2(celsius):
    return celsius * 9 / 5 + 3  # พิมพ์ผิด (32 กลายเป็น 3) เพราะคัดลอกสูตรไปเขียนซ้ำ

temp = 25
print(celsius_to_fahrenheit_v1(temp))
print(celsius_to_fahrenheit_v2(temp))

77.0
48.0


**Output:**
```
77.0
48.0
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 7 | `temp = 25` | temp=25 | กำหนดค่าตั้งต้น |
| 8 | เรียก `celsius_to_fahrenheit_v1(25)` | celsius=25 | เข้าฟังก์ชันเวอร์ชันที่ถูกต้อง |
| 2 | `return celsius * 9 / 5 + 32` | คืนค่า 77.0 | สูตรแปลงองศาที่ถูกต้อง |
| 8 | `print(...)` | แสดงผล 77.0 | ผลลัพธ์ที่คาดหวัง |
| 9 | เรียก `celsius_to_fahrenheit_v2(25)` | celsius=25 | เข้าฟังก์ชันเวอร์ชันที่คัดลอกสูตรมาเขียนใหม่ |
| 5 | `return celsius * 9 / 5 + 3` | คืนค่า 48.0 | สูตรผิดเพราะพิมพ์เลข 32 เป็น 3 ตอนคัดลอก |
| 9 | `print(...)` | แสดงผล 48.0 | ผลลัพธ์ผิดพลาด ทั้งที่ตั้งใจให้เหมือนกับฟังก์ชันแรก |

**คำอธิบายภาพรวม:** บรรทัด 1–2 และ 4–5 เป็นฟังก์ชันสองตัวที่ผู้เขียนตั้งใจให้ทำงานเหมือนกัน แต่ `celsius_to_fahrenheit_v2`
เกิดจากการคัดลอกสูตรของฟังก์ชันแรกมาพิมพ์ใหม่แทนที่จะเรียกใช้ฟังก์ชันเดิมซ้ำ บรรทัด 8–9 เรียกทั้งสองฟังก์ชันด้วยอินพุตเดียวกัน (25)
แต่ได้ผลลัพธ์ต่างกัน (77.0 กับ 48.0) เพราะพิมพ์เลข 32 ผิดเป็น 3 ระหว่างคัดลอก ตัวอย่างนี้แสดงความเสี่ยงของการไม่ทำ DRY:
เมื่อตรรกะเดียวกันถูกเขียนซ้ำหลายแห่ง การพิมพ์ผิดหรือแก้ไขในจุดหนึ่งจะไม่ส่งผลไปยังจุดอื่นโดยอัตโนมัติ ทำให้ตรรกะของโปรแกรมไม่สอดคล้องกัน

### แหล่งอ้างอิง
- W3Schools — Python Functions (https://www.w3schools.com/python/python_functions.asp): อธิบายว่าฟังก์ชันช่วยหลีกเลี่ยงการเขียนโค้ดซ้ำ
- TutorialsPoint — Python Functions (https://www.tutorialspoint.com/python/python_functions.htm)
- GeeksforGeeks — Python Functions, หัวข้อ Defining a Function (https://www.geeksforgeeks.org/python/python-functions/)



---

## 2: Function Definition กับ Function Call

### แนวคิด

การใช้งานฟังก์ชันแบ่งออกเป็นสองขั้นตอนที่แยกจากกันอย่างชัดเจน ขั้นตอนแรกคือ **การนิยามฟังก์ชัน (function definition)**
คือการบอกให้โปรแกรมรู้จักว่าฟังก์ชันนี้ชื่ออะไร รับข้อมูลอะไรเข้ามา และมีคำสั่งอะไรอยู่ภายในบ้าง ขั้นตอนนี้เป็นเพียงการ "ประกาศ" ไว้เท่านั้น
คำสั่งภายในฟังก์ชันจะยังไม่ถูกรันจนกว่าจะถึงขั้นตอนที่สอง คือ **การเรียกใช้ฟังก์ชัน (function call)**
ซึ่งเป็นจุดที่โปรแกรมกระโดดเข้าไปรันคำสั่งภายในฟังก์ชันนั้นจริง ๆ ฟังก์ชันหนึ่งสามารถถูกนิยามเพียงครั้งเดียว แต่ถูกเรียกใช้ได้หลายครั้งตามต้องการ
ความเข้าใจผิดที่พบบ่อยของผู้เริ่มต้นคือคิดว่าเพียงแค่นิยามฟังก์ชันไว้ โปรแกรมก็จะรันคำสั่งภายในให้โดยอัตโนมัติ ซึ่งไม่เป็นความจริง

ในภาษา Python การนิยามฟังก์ชันใช้คำสั่ง `def ชื่อฟังก์ชัน():` ตามด้วยกลุ่มคำสั่งที่เยื้องเข้าไป (indent)
ส่วนการเรียกใช้ทำได้โดยเขียนชื่อฟังก์ชันตามด้วยวงเล็บ `()` ดังตัวอย่างต่อไปนี้

### ตัวอย่าง Python

#### ตัวอย่างที่ 2.1 — แนวคิดหลัก: นิยามครั้งเดียว เรียกใช้ได้หลายครั้ง

In [4]:
def greet():
    print("สวัสดี")

print("ก่อนเรียกฟังก์ชัน")
greet()
greet()
print("หลังเรียกฟังก์ชัน")

ก่อนเรียกฟังก์ชัน
สวัสดี
สวัสดี
หลังเรียกฟังก์ชัน


**Output:**
```
ก่อนเรียกฟังก์ชัน
สวัสดี
สวัสดี
หลังเรียกฟังก์ชัน
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1–2 | นิยามฟังก์ชัน `greet` | (ยังไม่ทำงาน) | เพียงบันทึกนิยามไว้ ยังไม่มีการรันโค้ดข้างในบรรทัด 2 |
| 4 | `print("ก่อนเรียกฟังก์ชัน")` | แสดงผล | ทำงานตามลำดับปกติของโปรแกรม |
| 5 | เรียก `greet()` ครั้งที่ 1 | เข้าสู่ scope ของ greet | เริ่มรันโค้ดในฟังก์ชัน |
| 2 | `print("สวัสดี")` | แสดงผล | พิมพ์จากภายในฟังก์ชัน |
| 6 | เรียก `greet()` ครั้งที่ 2 | เข้าสู่ scope ของ greet อีกครั้ง | เรียกซ้ำด้วยนิยามเดิม ได้ผลเหมือนเดิม |
| 2 | `print("สวัสดี")` | แสดงผล | พิมพ์ซ้ำ |
| 7 | `print("หลังเรียกฟังก์ชัน")` | แสดงผล | พิมพ์ข้อความหลังเรียกฟังก์ชันครบ 2 ครั้ง |

**คำอธิบายภาพรวม:** บรรทัด 1–2 เป็นเพียงการนิยามฟังก์ชัน (definition) ณ จุดนี้ Python ยังไม่รันคำสั่ง `print` ที่อยู่ภายใน
มีเพียงการสร้างฟังก์ชันไว้ใช้งานเท่านั้น บรรทัด 5 และ 6 คือการเรียกใช้ (call) ฟังก์ชันเดิมสองครั้ง ทำให้โค้ดในบรรทัด 2 ถูกรันจริงสองครั้ง
แสดงให้เห็นว่า "การนิยาม" กับ "การเรียกใช้" เป็นคนละขั้นตอนกัน นิยามเพียงครั้งเดียวแต่เรียกใช้ได้หลายครั้งตามต้องการ

#### ตัวอย่างที่ 2.2 — บริบทอื่นที่ต่างโดเมน: เรียกฟังก์ชันเดิมซ้ำผ่านลูปในงานประมวลผลตัวเลข

In [5]:
def square(number):
    return number * number

numbers = [2, 4, 6]
for n in numbers:
    print(square(n))

4
16
36


**Output:**
```
4
16
36
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1–2 | นิยามฟังก์ชัน `square` | (ยังไม่ทำงาน) | นิยามเพียงครั้งเดียวก่อนเริ่มลูป |
| 4 | `numbers = [2, 4, 6]` | numbers=[2, 4, 6] | สร้างลิสต์ตัวเลขที่จะวนใช้ |
| 5 | `for n in numbers:` (รอบที่ 1) | n=2 | เริ่มวนลูป |
| 6 | เรียก `square(2)` | number=2 | เข้าสู่ฟังก์ชัน (call) ครั้งที่ 1 |
| 2 | `return number * number` | คืนค่า 4 | คำนวณจากนิยามเดิมที่มีอยู่ที่เดียว |
| 6 | `print(...)` | แสดงผล 4 | พิมพ์ผลลัพธ์ครั้งที่ 1 |
| 5 | `for n in numbers:` (รอบที่ 2) | n=4 | วนรอบถัดไป |
| 6 | เรียก `square(4)` | number=4 | เข้าสู่ฟังก์ชัน ครั้งที่ 2 |
| 2 | `return number * number` | คืนค่า 16 | ใช้นิยามเดิมซ้ำ |
| 6 | `print(...)` | แสดงผล 16 | พิมพ์ผลลัพธ์ครั้งที่ 2 |
| 5 | `for n in numbers:` (รอบที่ 3) | n=6 | วนรอบสุดท้าย |
| 6 | เรียก `square(6)` | number=6 | เข้าสู่ฟังก์ชัน ครั้งที่ 3 |
| 2 | `return number * number` | คืนค่า 36 | ใช้นิยามเดิมซ้ำ |
| 6 | `print(...)` | แสดงผล 36 | พิมพ์ผลลัพธ์ครั้งที่ 3 |

**คำอธิบายภาพรวม:** บรรทัด 1–2 นิยามฟังก์ชัน `square` เพียงครั้งเดียว เช่นเดียวกับหลักการในตัวอย่างที่ 2.1 บรรทัด 4–6 แสดงให้เห็นว่าการเรียกใช้ฟังก์ชัน
(call) ไม่จำเป็นต้องเขียนชื่อฟังก์ชันซ้ำด้วยมือทีละครั้งเสมอไป แต่สามารถเรียกซ้ำโดยอัตโนมัติผ่านคำสั่งวนลูป (`for`) ได้เช่นกัน โดยฟังก์ชันเดิมที่ถูกนิยามไว้
เพียงบรรทัด 1–2 ถูกเรียกใช้ 3 ครั้งด้วยค่าต่างกันจากลิสต์ `numbers` สะท้อนแนวคิดเดียวกับตัวอย่างที่ 2.1 ว่า "นิยามครั้งเดียว เรียกใช้ได้หลายครั้ง"
แต่อยู่ในบริบทของการประมวลผลตัวเลขในลิสต์แทนข้อความทักทาย

#### ตัวอย่างที่ 2.3 — ข้อผิดพลาดที่พบบ่อย: นิยามฟังก์ชันแล้วลืมเรียกใช้

In [6]:
def show_message():
    print("ทำงานแล้ว")

print("เริ่มโปรแกรม")
# ลืมเรียกใช้ฟังก์ชัน show_message() ตรงนี้
print("จบโปรแกรม")

เริ่มโปรแกรม
จบโปรแกรม


**Output:**
```
เริ่มโปรแกรม
จบโปรแกรม
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1–2 | นิยามฟังก์ชัน `show_message` | (ยังไม่ทำงาน) | สร้างฟังก์ชันไว้ ยังไม่ถูกเรียกใช้ |
| 4 | `print("เริ่มโปรแกรม")` | แสดงผล | ทำงานตามลำดับ |
| 5 | (ไม่มีคำสั่งเรียกฟังก์ชัน) | — | ไม่มี `show_message()` ปรากฏอยู่ในโปรแกรมเลย |
| 6 | `print("จบโปรแกรม")` | แสดงผล | โปรแกรมจบการทำงานโดยไม่เคยรันโค้ดในฟังก์ชัน |

**คำอธิบายภาพรวม:** บรรทัด 1–2 นิยามฟังก์ชัน `show_message` ไว้ แต่ตลอดทั้งโปรแกรมไม่มีบรรทัดใดเรียกใช้ฟังก์ชันนี้เลย
(ไม่มี `show_message()` ปรากฏ) ผลลัพธ์จึงไม่มีข้อความ "ทำงานแล้ว" ปรากฏออกมา ทั้งที่ฟังก์ชันถูกนิยามไว้แล้วก็ตาม
ตัวอย่างนี้สะท้อนข้อผิดพลาดที่พบบ่อยของผู้เริ่มต้น คือเข้าใจผิดว่าการนิยามฟังก์ชัน (`def`) เพียงอย่างเดียวจะทำให้โค้ดภายในทำงานทันที
ทั้งที่ในความเป็นจริงต้องมีการเรียกใช้ (call) ด้วยชื่อฟังก์ชันตามด้วยวงเล็บเสมอ โปรแกรมจึงจะรันคำสั่งภายในฟังก์ชันนั้น

### แหล่งอ้างอิง
- W3Schools — Python Functions (https://www.w3schools.com/python/python_functions.asp): "To call a function, write its name followed by parentheses."
- TutorialsPoint — Python Functions (https://www.tutorialspoint.com/python/python_functions.htm)
- GeeksforGeeks — Python Functions, หัวข้อ Defining a Function / Calling a Function (https://www.geeksforgeeks.org/python/python-functions/)

---

## 3: Parameter กับ Argument

### แนวคิด

เมื่อออกแบบฟังก์ชันให้รับข้อมูลเข้ามาประมวลผล จำเป็นต้องเข้าใจคำศัพท์สองคำที่มักถูกใช้ปะปนกัน คือ **พารามิเตอร์ (parameter)**
และ **อาร์กิวเมนต์ (argument)** พารามิเตอร์คือชื่อตัวแปรที่ถูกระบุไว้ในวงเล็บตอน**นิยาม**ฟังก์ชัน ทำหน้าที่เป็นเพียง "ช่องว่าง"
ที่รอรับค่าเข้ามา ยังไม่มีค่าจริงอยู่ในตัว ส่วนอาร์กิวเมนต์คือค่าจริงที่ถูกส่งเข้าไปตอน**เรียกใช้**ฟังก์ชัน ซึ่งจะถูกนำไปผูก (bind)
เข้ากับพารามิเตอร์ตามลำดับตำแหน่งที่ปรากฏ วิธีการส่งอาร์กิวเมนต์ตามลำดับตำแหน่งเช่นนี้เรียกว่า
**อาร์กิวเมนต์ตามตำแหน่ง (positional argument)** ซึ่งอาร์กิวเมนต์ตัวแรกจะถูกผูกกับพารามิเตอร์ตัวแรกเสมอ
ตัวที่สองผูกกับตัวที่สอง เรียงตามลำดับไปเรื่อย ๆ โดยภาษาโปรแกรมส่วนใหญ่จะไม่ตรวจสอบว่าค่าที่ส่งเข้ามาตรงกับความหมายที่ต้องการหรือไม่
ทำให้การส่งอาร์กิวเมนต์ผิดลำดับเป็นข้อผิดพลาดที่พบได้บ่อยและตรวจจับได้ยาก

ในภาษา Python พารามิเตอร์ถูกระบุไว้ในวงเล็บของ `def` เช่น `def introduce(name, age):` ส่วนอาร์กิวเมนต์คือค่าที่ส่งเข้าไปตอนเรียก
เช่น `introduce("Aom", 20)` ซึ่ง `"Aom"` จะถูกผูกกับ `name` และ `20` จะถูกผูกกับ `age` ตามลำดับตำแหน่ง

### ตัวอย่าง Python

#### ตัวอย่างที่ 3.1 — แนวคิดหลัก: พารามิเตอร์รับค่าจากอาร์กิวเมนต์ตามลำดับตำแหน่ง

In [7]:
def introduce(name, age):
    print(f"ฉันชื่อ {name} อายุ {age} ปี")

introduce("Aom", 20)
introduce("Beam", 22)

ฉันชื่อ Aom อายุ 20 ปี
ฉันชื่อ Beam อายุ 22 ปี


**Output:**
```
ฉันชื่อ Aom อายุ 20 ปี
ฉันชื่อ Beam อายุ 22 ปี
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1 | นิยามฟังก์ชัน `introduce(name, age)` | name, age เป็นพารามิเตอร์ | กำหนดชื่อตัวแปรที่จะรับค่า ยังไม่มีค่าจริง |
| 4 | เรียก `introduce("Aom", 20)` | name="Aom", age=20 | ส่งอาร์กิวเมนต์ "Aom" และ 20 เข้าไปผูกกับพารามิเตอร์ตามลำดับ |
| 2 | `print(f"...")` | แสดงผล ฉันชื่อ Aom อายุ 20 ปี | ใช้ค่าของพารามิเตอร์ในการพิมพ์ |
| 5 | เรียก `introduce("Beam", 22)` | name="Beam", age=22 | เรียกซ้ำด้วยอาร์กิวเมนต์ชุดใหม่ |
| 2 | `print(f"...")` | แสดงผล ฉันชื่อ Beam อายุ 22 ปี | พารามิเตอร์ชุดเดิม รับค่าใหม่ทุกครั้งที่ถูกเรียก |

**คำอธิบายภาพรวม:** บรรทัด 1 นิยามฟังก์ชันโดยระบุพารามิเตอร์ `name` และ `age` ซึ่งเป็นเพียงชื่อตัวแปรที่รอรับค่า ยังไม่ใช่ค่าจริง
บรรทัด 4 และ 5 คือการเรียกใช้ฟังก์ชันพร้อมส่งอาร์กิวเมนต์ ได้แก่ค่าจริงเช่น `"Aom"` และ `20` เข้าไปผูกกับพารามิเตอร์ตามลำดับตำแหน่ง
ทำให้ฟังก์ชันเดียวกันสามารถประมวลผลข้อมูลที่แตกต่างกันได้ในแต่ละครั้งที่ถูกเรียก

#### ตัวอย่างที่ 3.2 — บริบทอื่นที่ต่างโดเมน: คำนวณราคารวมด้วยอาร์กิวเมนต์ตามตำแหน่งและแบบระบุชื่อ

In [8]:
def calculate_total(price, quantity):
    return price * quantity

print(calculate_total(50, 3))
print(calculate_total(quantity=3, price=50))

150
150


**Output:**
```
150
150
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1 | นิยามฟังก์ชัน `calculate_total(price, quantity)` | price, quantity เป็นพารามิเตอร์ | กำหนดชื่อตัวแปรที่จะรับค่า ยังไม่มีค่าจริง |
| 4 | เรียก `calculate_total(50, 3)` | price=50, quantity=3 | ส่งอาร์กิวเมนต์ตามตำแหน่ง (positional) 50 ผูกกับ price, 3 ผูกกับ quantity |
| 2 | `return price * quantity` | คืนค่า 150 | คำนวณราคารวมสินค้า |
| 4 | `print(...)` | แสดงผล 150 | ผลลัพธ์จากการเรียกแบบ positional |
| 5 | เรียก `calculate_total(quantity=3, price=50)` | price=50, quantity=3 | ส่งอาร์กิวเมนต์แบบระบุชื่อ (keyword argument) ลำดับที่เขียนสลับกันได้เพราะระบุชื่อพารามิเตอร์กำกับไว้ |
| 2 | `return price * quantity` | คืนค่า 150 | ได้ผลลัพธ์เดียวกับการเรียกแบบ positional |
| 5 | `print(...)` | แสดงผล 150 | ผลลัพธ์จากการเรียกแบบ keyword argument |

**คำอธิบายภาพรวม:** บรรทัด 1–2 นิยามฟังก์ชัน `calculate_total` ในบริบทการคำนวณราคารวมสินค้า ต่างจากตัวอย่างที่ 3.1 ที่เป็นการแนะนำตัวบุคคล
บรรทัด 4 เรียกฟังก์ชันด้วยอาร์กิวเมนต์ตามตำแหน่ง (positional argument) เช่นเดียวกับตัวอย่างที่ 3.1 ส่วนบรรทัด 5 สาธิตอาร์กิวเมนต์แบบระบุชื่อ
(keyword argument) ซึ่งระบุชื่อพารามิเตอร์กำกับไว้ตอนเรียกใช้ ทำให้สลับลำดับการเขียนอาร์กิวเมนต์ได้โดยไม่เปลี่ยนความหมาย เพราะ Python
ผูกค่าตามชื่อพารามิเตอร์แทนที่จะผูกตามตำแหน่งเมื่อระบุชื่อไว้ชัดเจน ต่างจากตัวอย่างที่ 3.3 (ถัดไป) ที่แสดงว่าการส่งอาร์กิวเมนต์ตามตำแหน่งผิดลำดับ
จะทำให้ค่าผูกผิดพารามิเตอร์โดยไม่มี error เตือน

#### ตัวอย่างที่ 3.3 — ข้อผิดพลาดที่พบบ่อย: ส่งอาร์กิวเมนต์ผิดลำดับ

In [9]:
def introduce(name, age):
    print(f"ฉันชื่อ {name} อายุ {age} ปี")

introduce("Aom", 20)   # ลำดับถูกต้อง: name, age ตรงกับพารามิเตอร์
introduce(20, "Aom")   # ลำดับผิด: สลับตำแหน่งอาร์กิวเมนต์

ฉันชื่อ Aom อายุ 20 ปี
ฉันชื่อ 20 อายุ Aom ปี


**Output:**
```
ฉันชื่อ Aom อายุ 20 ปี
ฉันชื่อ 20 อายุ Aom ปี
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 4 | เรียก `introduce("Aom", 20)` | name="Aom", age=20 | อาร์กิวเมนต์ตามตำแหน่งตรงกับความหมายของพารามิเตอร์ |
| 2 | `print(f"...")` | แสดงผล ฉันชื่อ Aom อายุ 20 ปี | ผลลัพธ์ถูกต้องตามที่ตั้งใจ |
| 5 | เรียก `introduce(20, "Aom")` | name=20, age="Aom" | สลับตำแหน่งอาร์กิวเมนต์ Python ผูกค่าตามลำดับโดยไม่ตรวจชนิดข้อมูล |
| 2 | `print(f"...")` | แสดงผล ฉันชื่อ 20 อายุ Aom ปี | ไม่มี error แต่ความหมายผิด เพราะพารามิเตอร์รับค่าผิดตัว |

**คำอธิบายภาพรวม:** บรรทัด 4 เรียกฟังก์ชันด้วยอาร์กิวเมนต์ตามตำแหน่งที่เรียงลำดับถูกต้องตรงกับพารามิเตอร์ `name` และ `age`
บรรทัด 5 สาธิตข้อผิดพลาดที่พบบ่อย คือสลับลำดับอาร์กิวเมนต์ ซึ่ง Python จะไม่แจ้งข้อผิดพลาดใด ๆ เพราะภาษาไม่ได้ตรวจสอบชนิดข้อมูลของพารามิเตอร์
แต่จะผูกค่าตามตำแหน่งเสมอ ผลลัพธ์จึงผิดเพี้ยนไปจากที่ตั้งใจโดยที่โปรแกรมยังทำงานได้ตามปกติ (ไม่มี error หยุดโปรแกรม)
จึงเป็นข้อผิดพลาดเชิงตรรกะที่ตรวจพบได้ยากกว่าข้อผิดพลาดที่ทำให้โปรแกรมหยุดทำงาน


### แหล่งอ้างอิง
- W3Schools — Python Function Arguments (https://www.w3schools.com/python/python_arguments.asp): แยกความหมาย parameter (ตัวแปรในวงเล็บตอนนิยาม) กับ argument (ค่าจริงตอนเรียก) ไว้ชัดเจน
- TutorialsPoint — Positional and Keyword Argument (https://www.tutorialspoint.com/keyword-and-positional-argument-in-python): อธิบายอาร์กิวเมนต์ตามตำแหน่งว่าต้องเรียงลำดับให้ตรงกับพารามิเตอร์
- GeeksforGeeks — Python Functions, หัวข้อ Function Arguments / Positional Arguments (https://www.geeksforgeeks.org/python/python-functions/)


---

## 4: Return Value กับ Procedure (Void Function)

### แนวคิด

ฟังก์ชันแบ่งออกได้เป็นสองลักษณะตามพฤติกรรมการส่งผลลัพธ์กลับ ลักษณะแรกคือฟังก์ชันที่มี**ค่าที่ส่งกลับ (return value)**
คือฟังก์ชันที่ประมวลผลบางอย่างแล้วส่งผลลัพธ์กลับไปให้จุดที่เรียกใช้ นำไปเก็บในตัวแปรหรือใช้ในนิพจน์ต่อได้ทันที
ทำได้ด้วยคำสั่ง `return` ซึ่งเมื่อถูกรันจะหยุดการทำงานของฟังก์ชันทันทีและส่งค่าที่ระบุกลับไป ลักษณะที่สองคือฟังก์ชันแบบ
**procedure** (บางครั้งเรียกว่า **void function**) คือฟังก์ชันที่ทำงานบางอย่าง เช่น พิมพ์ข้อความออกหน้าจอ แต่ไม่มีคำสั่ง `return`
ที่ส่งค่ากลับมาให้ใช้งานต่อ ฟังก์ชันประเภทนี้ถูกเรียกใช้เพื่อ "ผลข้างเคียง" (side effect) ของมัน เช่นการแสดงผล ไม่ใช่เพื่อค่าที่ได้กลับมา

ในภาษา Python หากฟังก์ชันใดไม่มีคำสั่ง `return` หรือมี `return` โดยไม่ระบุค่า ฟังก์ชันนั้นจะคืนค่าพิเศษที่ชื่อ `None` กลับไปโดยอัตโนมัติเสมอ
ซึ่งเป็นพฤติกรรมสำคัญที่ต้องเข้าใจ เพราะการนำผลลัพธ์จากฟังก์ชันแบบ procedure ไปใช้งานต่อราวกับว่ามันมีค่าที่ส่งกลับ จะได้ `None` แทนค่าที่ต้องการ

### ตัวอย่าง Python

#### ตัวอย่างที่ 4.1 — แนวคิดหลัก: เปรียบเทียบฟังก์ชันที่มี return กับฟังก์ชันแบบ procedure

In [10]:
def add(x, y):
    return x + y

def show_greeting(name):
    print(f"สวัสดี {name}")

total = add(3, 5)
result = show_greeting("Aom")
print(total)
print(result)

สวัสดี Aom
8
None


**Output:**
```
สวัสดี Aom
8
None
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 7 | เรียก `add(3, 5)` | x=3, y=5 | เข้าฟังก์ชัน add |
| 2 | `return x + y` | คืนค่า 8 | ส่งค่ากลับไปเก็บในตัวแปร total |
| 7 | `total = ...` | total=8 | ตัวแปร total เก็บค่าที่ฟังก์ชันส่งกลับ |
| 8 | เรียก `show_greeting("Aom")` | name="Aom" | เข้าฟังก์ชัน show_greeting |
| 5 | `print(f"สวัสดี {name}")` | แสดงผล สวัสดี Aom | ฟังก์ชันนี้พิมพ์ข้อความเป็นผลข้างเคียง (side effect) ไม่มีคำสั่ง return |
| 8 | `result = ...` | result=None | ไม่มี return ในฟังก์ชัน จึงได้ค่า None โดยปริยาย |
| 9 | `print(total)` | แสดงผล 8 | ค่าที่ได้จากฟังก์ชันที่มี return |
| 10 | `print(result)` | แสดงผล None | ค่าที่ได้จากฟังก์ชันแบบ procedure |

**คำอธิบายภาพรวม:** บรรทัด 1–2 เป็นฟังก์ชันที่มีค่าที่ส่งกลับ คือ `add` ซึ่งคำนวณแล้วส่งผลลัพธ์กลับให้ผู้เรียกนำไปใช้ต่อ
ส่วนบรรทัด 4–5 เป็นฟังก์ชันแบบ procedure คือ `show_greeting` ที่ทำหน้าที่พิมพ์ข้อความเป็นผลข้างเคียงเท่านั้น ไม่มีคำสั่ง `return`
บรรทัด 7–10 แสดงผลต่างที่ชัดเจน: การเก็บผลจาก `add` ได้ค่า 8 มาใช้งานต่อได้จริง ในขณะที่การเก็บผลจาก `show_greeting`
ได้ค่า `None` เพราะฟังก์ชันที่ไม่มี `return` จะคืนค่า `None` โดยอัตโนมัติเสมอ

#### ตัวอย่างที่ 4.2 — บริบทอื่นที่ต่างโดเมน: return value กับ procedure ในงานประมวลผลลิสต์

In [11]:
def find_max(numbers):
    return max(numbers)

def print_list(items):
    for item in items:
        print(item)

scores = [88, 92, 79]
highest = find_max(scores)
outcome = print_list(scores)
print(highest)
print(outcome)

88
92
79
92
None


**Output:**
```
88
92
79
92
None
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 8 | `scores = [88, 92, 79]` | scores=[88, 92, 79] | สร้างลิสต์คะแนน |
| 9 | เรียก `find_max(scores)` | numbers=[88, 92, 79] | เข้าฟังก์ชันที่มีค่าที่ส่งกลับ |
| 2 | `return max(numbers)` | คืนค่า 92 | คำนวณค่าสูงสุดแล้วส่งกลับ |
| 9 | `highest = ...` | highest=92 | เก็บค่าที่ฟังก์ชันส่งกลับ |
| 10 | เรียก `print_list(scores)` | items=[88, 92, 79] | เข้าฟังก์ชันแบบ procedure |
| 5–6 | `for item in items: / print(item)` | แสดงผล 88, 92, 79 ตามลำดับ | พิมพ์ค่าทีละตัวเป็นผลข้างเคียง ไม่มีคำสั่ง return |
| 10 | `outcome = ...` | outcome=None | ฟังก์ชันไม่มี return จึงได้ None โดยปริยาย |
| 11 | `print(highest)` | แสดงผล 92 | ค่าที่ได้จากฟังก์ชันที่มีค่าที่ส่งกลับ |
| 12 | `print(outcome)` | แสดงผล None | ค่าที่ได้จากฟังก์ชันแบบ procedure |

**คำอธิบายภาพรวม:** บรรทัด 1–2 เป็นฟังก์ชันที่มีค่าที่ส่งกลับ คือ `find_max` ซึ่งคำนวณค่าสูงสุดของลิสต์แล้วส่งผลลัพธ์กลับให้นำไปใช้ต่อ
บรรทัด 4–6 เป็นฟังก์ชันแบบ procedure ชื่อ `print_list` ที่วนพิมพ์สมาชิกแต่ละตัวในลิสต์เป็นผลข้างเคียง (side effect) ไม่มีคำสั่ง `return`
บรรทัด 8–12 ทดสอบทั้งสองฟังก์ชันกับลิสต์คะแนนเดียวกัน (`scores`) แสดงหลักการเดียวกับตัวอย่างที่ 4.1 แต่อยู่ในบริบทการประมวลผลข้อมูลแบบลิสต์
แทนตัวเลขเดี่ยวและข้อความทักทาย ยืนยันอีกครั้งว่าฟังก์ชันแบบ procedure จะคืนค่า `None` เสมอเมื่อนำไปเก็บไว้ในตัวแปร

#### ตัวอย่างที่ 4.3 — ข้อผิดพลาดที่พบบ่อย: เข้าใจผิดว่าฟังก์ชันแบบ procedure ส่งค่ากลับมาด้วย

In [12]:
def show_greeting(name):
    print(f"สวัสดี {name}")

message = show_greeting("Beam")

try:
    print("ข้อความคือ: " + message)
except TypeError as error:
    print(f"เกิดข้อผิดพลาด: {error}")

สวัสดี Beam
เกิดข้อผิดพลาด: can only concatenate str (not "NoneType") to str


**Output:**
```
สวัสดี Beam
เกิดข้อผิดพลาด: can only concatenate str (not "NoneType") to str
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 4 | เรียก `show_greeting("Beam")` | name="Beam" | เข้าฟังก์ชัน |
| 2 | `print(f"สวัสดี {name}")` | แสดงผล สวัสดี Beam | พิมพ์ข้อความเป็นผลข้างเคียง |
| 4 | `message = ...` | message=None | ฟังก์ชันไม่มี return จึงได้ None |
| 7 | `"ข้อความคือ: " + message` | พยายามต่อ string กับ None | เกิด TypeError เพราะบวก str กับ NoneType ไม่ได้ |
| 8–9 | `except TypeError as error` | แสดงผล เกิดข้อผิดพลาด: can only concatenate str (not "NoneType") to str | ดักข้อผิดพลาดและพิมพ์ข้อความแทนการหยุดโปรแกรมกะทันหัน |

**คำอธิบายภาพรวม:** บรรทัด 4 เก็บผลลัพธ์จาก `show_greeting` ไว้ในตัวแปร `message` ซึ่งมีค่าเป็น `None` เพราะฟังก์ชันนี้เป็น procedure
ที่ไม่มี `return` บรรทัด 6–9 สาธิตข้อผิดพลาดที่พบบ่อยเมื่อผู้เขียนโปรแกรมเข้าใจผิดว่าฟังก์ชันที่พิมพ์ข้อความออกหน้าจอ (`print`)
จะ "ส่งค่ากลับ" ข้อความนั้นให้ใช้งานต่อได้ด้วย เมื่อพยายามนำ `message` ไปต่อกับ string อื่นจึงเกิด `TypeError` เพราะ `message`
มีค่าเป็น `None` ไม่ใช่ string จริง ตัวอย่างนี้ย้ำความแตกต่างระหว่างฟังก์ชันที่มีค่าที่ส่งกลับกับฟังก์ชันแบบ procedure ให้ชัดเจนยิ่งขึ้น


### แหล่งอ้างอิง
- W3Schools — Python Functions (https://www.w3schools.com/python/python_functions.asp): "The `return` statement... stops the execution of the function, and 'returns' the result."
- TutorialsPoint — Python Functions, หัวข้อ return statement (https://www.tutorialspoint.com/python/python_functions.htm)
- GeeksforGeeks — Python Functions, หัวข้อ Return Statement (https://www.geeksforgeeks.org/python/python-functions/): "Return is used to end a function and send a value back to the caller."


---

## 5: ขอบเขตตัวแปร (Scope) — ตัวแปรเฉพาะที่ (Local) กับตัวแปรส่วนกลาง (Global)

### แนวคิด

**ขอบเขต (scope)** ของตัวแปร หมายถึงบริเวณของโปรแกรมที่ตัวแปรนั้นสามารถถูกอ้างถึงและใช้งานได้ ตัวแปรที่ถูกสร้างขึ้นภายในฟังก์ชัน
เรียกว่า **ตัวแปรเฉพาะที่ (local variable)** จะมีขอบเขตจำกัดอยู่เฉพาะภายในฟังก์ชันนั้นเท่านั้น เมื่อฟังก์ชันทำงานเสร็จสิ้น
ตัวแปรเฉพาะที่จะถูกทำลายไป และไม่สามารถอ้างถึงได้จากภายนอกฟังก์ชัน ในทางกลับกัน ตัวแปรที่ถูกสร้างขึ้นภายนอกฟังก์ชันทั้งหมด (ในระดับบนสุดของโปรแกรม)
เรียกว่า **ตัวแปรส่วนกลาง (global variable)** จะมีขอบเขตครอบคลุมทั้งโปรแกรม สามารถอ่านค่าได้จากทุกจุด รวมถึงจากภายในฟังก์ชันต่าง ๆ ด้วย

อย่างไรก็ตาม การ**อ่าน**ค่าตัวแปรส่วนกลางจากภายในฟังก์ชันทำได้โดยตรงโดยไม่ต้องประกาศอะไรเพิ่มเติม แต่การ**แก้ไข**ค่าตัวแปรส่วนกลาง
จากภายในฟังก์ชันนั้น จำเป็นต้องประกาศชื่อตัวแปรด้วยคำสั่งพิเศษก่อน (ในภาษา Python คือคำสั่ง `global`) มิฉะนั้นภาษาโปรแกรมจะตีความว่ากำลังสร้าง
ตัวแปรเฉพาะที่ตัวใหม่ขึ้นมาแทน ซึ่งเป็นแหล่งความเข้าใจผิดที่พบบ่อยที่สุดข้อหนึ่งของผู้เริ่มต้นเขียนโปรแกรม

### ตัวอย่าง Python

#### ตัวอย่างที่ 5.1 — แนวคิดหลัก: อ่านค่า global ได้อิสระ แต่ต้องประกาศ `global` ก่อนจึงจะแก้ไขได้

In [13]:
balance = 100

def deposit(amount):
    global balance
    balance = balance + amount

def show_balance():
    print("ยอดคงเหลือ:", balance)

show_balance()
deposit(50)
show_balance()

ยอดคงเหลือ: 100
ยอดคงเหลือ: 150


**Output:**
```
ยอดคงเหลือ: 100
ยอดคงเหลือ: 150
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1 | `balance = 100` | balance=100 (global) | สร้างตัวแปรส่วนกลาง |
| 10 | เรียก `show_balance()` ครั้งที่ 1 | อ่านค่า balance จาก global scope | ฟังก์ชันอ่านตัวแปรส่วนกลางได้โดยไม่ต้องประกาศ global |
| 8 | `print("ยอดคงเหลือ:", balance)` | แสดงผล ยอดคงเหลือ: 100 | ค่าปัจจุบันของ balance |
| 11 | เรียก `deposit(50)` | amount=50 | เข้าฟังก์ชัน deposit |
| 4 | `global balance` | ประกาศว่าจะใช้ตัวแปร balance จาก global scope | หากไม่มีบรรทัดนี้ บรรทัดถัดไปจะสร้างตัวแปร local ใหม่แทน (ดูตัวอย่างที่ 5.3) |
| 5 | `balance = balance + amount` | balance=150 (global ถูกแก้ไข) | แก้ไขค่าตัวแปรส่วนกลางจริง เพราะมี global กำกับไว้ |
| 12 | เรียก `show_balance()` ครั้งที่ 2 | อ่านค่า balance ล่าสุด | ยืนยันว่าค่า global ถูกเปลี่ยนแปลงจริง |
| 8 | `print(...)` | แสดงผล ยอดคงเหลือ: 150 | ค่าที่เปลี่ยนแปลงคงอยู่ข้ามการเรียกฟังก์ชัน |

**คำอธิบายภาพรวม:** บรรทัด 1 สร้างตัวแปรส่วนกลาง `balance` บรรทัด 7–8 แสดงว่าฟังก์ชัน `show_balance` สามารถอ่านค่าตัวแปรส่วนกลางได้โดยตรง
โดยไม่ต้องประกาศคำสั่งใดเพิ่มเติม ส่วนบรรทัด 3–5 แสดงกรณีที่ต้องการแก้ไขค่าตัวแปรส่วนกลางจากภายในฟังก์ชัน ซึ่งจำเป็นต้องประกาศ `global balance`
ก่อนเสมอ มิฉะนั้น Python จะตีความว่า `balance` ในบรรทัด 5 เป็นตัวแปรเฉพาะที่ตัวใหม่ บรรทัด 10–12 ยืนยันว่าการแก้ไขผ่าน `deposit`
ส่งผลต่อค่า `balance` ในขอบเขตส่วนกลางจริง เพราะการเรียก `show_balance` ครั้งที่สองพิมพ์ค่าที่เปลี่ยนแปลงแล้ว

#### ตัวอย่างที่ 5.2 — บริบทอื่นที่ต่างโดเมน: scope ของตัวแปรข้อความในระบบสถานะเกม

In [14]:
def update_status(new_status):
    global game_status
    game_status = new_status

def show_status():
    print("สถานะเกม:", game_status)

game_status = "กำลังเล่น"
show_status()
update_status("จบเกม")
show_status()

สถานะเกม: กำลังเล่น
สถานะเกม: จบเกม


**Output:**
```
สถานะเกม: กำลังเล่น
สถานะเกม: จบเกม
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 8 | `game_status = "กำลังเล่น"` | game_status="กำลังเล่น" (global) | สร้างตัวแปรส่วนกลาง |
| 9 | เรียก `show_status()` ครั้งที่ 1 | อ่านค่า game_status จาก global scope | ฟังก์ชันอ่านตัวแปรส่วนกลางได้โดยไม่ต้องประกาศ global |
| 6 | `print("สถานะเกม:", game_status)` | แสดงผล สถานะเกม: กำลังเล่น | ค่าปัจจุบันของ game_status |
| 10 | เรียก `update_status("จบเกม")` | new_status="จบเกม" | เข้าฟังก์ชัน |
| 2 | `global game_status` | ประกาศว่าจะใช้ตัวแปร game_status จาก global scope | เช่นเดียวกับหลักการในตัวอย่างที่ 5.1 |
| 3 | `game_status = new_status` | game_status="จบเกม" (global ถูกแก้ไข) | แก้ไขค่าตัวแปรส่วนกลางจริง เพราะมี global กำกับไว้ |
| 11 | เรียก `show_status()` ครั้งที่ 2 | อ่านค่า game_status ล่าสุด | ยืนยันว่าค่า global ถูกเปลี่ยนแปลงจริง |
| 6 | `print(...)` | แสดงผล สถานะเกม: จบเกม | ค่าที่เปลี่ยนแปลงคงอยู่ข้ามการเรียกฟังก์ชัน |

**คำอธิบายภาพรวม:** บรรทัด 8 สร้างตัวแปรส่วนกลาง `game_status` เป็นข้อความ (string) แทนที่จะเป็นตัวเลขอย่างตัวอย่างที่ 5.1 (`balance`)
บรรทัด 5–6 แสดงว่าฟังก์ชัน `show_status` อ่านค่าตัวแปรส่วนกลางได้โดยตรง ส่วนบรรทัด 1–3 แสดงการแก้ไขค่าตัวแปรส่วนกลางจากภายในฟังก์ชัน
โดยต้องประกาศ `global game_status` ก่อนเสมอ เช่นเดียวกับหลักการในตัวอย่างที่ 5.1 บรรทัด 9–11 ยืนยันว่าการเรียก `update_status`
ส่งผลต่อค่า `game_status` ในขอบเขตส่วนกลางจริง แสดงให้เห็นว่าแนวคิดเรื่อง scope ใช้ได้กับตัวแปรชนิดข้อความเช่นเดียวกับตัวเลข
ไม่ได้จำกัดอยู่เฉพาะโดเมนการเงินเท่านั้น

#### ตัวอย่างที่ 5.3 — ข้อผิดพลาดที่พบบ่อย: ลืมประกาศ `global` ก่อนแก้ไขตัวแปรส่วนกลาง

In [15]:
balance = 100

def deposit_wrong(amount):
    balance = balance + amount  # ไม่ได้ประกาศ global ไว้ก่อน
    return balance

try:
    deposit_wrong(50)
except UnboundLocalError as error:
    print(f"เกิดข้อผิดพลาด: {error}")

print(balance)

เกิดข้อผิดพลาด: cannot access local variable 'balance' where it is not associated with a value
100


**Output:**
```
เกิดข้อผิดพลาด: cannot access local variable 'balance' where it is not associated with a value
100
```

**ตารางการเปลี่ยนแปลงค่า:**

| บรรทัดที่ | คำสั่งที่ทำงาน | ตัวแปร/ค่า | หมายเหตุ |
|---|---|---|---|
| 1 | `balance = 100` | balance=100 (global) | ตัวแปรส่วนกลาง |
| 8 | เรียก `deposit_wrong(50)` | amount=50 | เข้าฟังก์ชัน |
| 4 | `balance = balance + amount` | เกิด UnboundLocalError | มีการกำหนดค่า (assign) ให้ `balance` ภายในฟังก์ชันบรรทัด 4 ทำให้ Python ถือว่า `balance` เป็นตัวแปรเฉพาะที่ตลอดทั้งฟังก์ชัน แม้แต่ตอนอ่านค่าฝั่งขวาของเครื่องหมาย `=` ก็ยังไม่มีค่าให้อ่าน |
| 9–10 | `except UnboundLocalError as error` | แสดงผล เกิดข้อผิดพลาด: cannot access local variable 'balance' where it is not associated with a value | ดักข้อผิดพลาดไว้แทนที่จะปล่อยให้โปรแกรมหยุดทำงานกะทันหัน |
| 12 | `print(balance)` | แสดงผล 100 | ตัวแปร global balance ไม่ถูกแก้ไข เพราะฟังก์ชันไม่เคยไปถึงบรรทัด return |

**คำอธิบายภาพรวม:** บรรทัด 3–5 นิยามฟังก์ชัน `deposit_wrong` ที่ตั้งใจจะนำค่าตัวแปรส่วนกลาง `balance` มาบวกกับ `amount`
แล้วเก็บกลับที่ `balance` เดิม แต่เนื่องจากไม่ได้ประกาศ `global balance` ก่อน (ต่างจากตัวอย่างที่ 5.1) Python จะตีความว่าบรรทัด 4
กำลังสร้างตัวแปรเฉพาะที่ชื่อ `balance` ขึ้นใหม่ทั้งฟังก์ชัน ทำให้เมื่ออ่านค่า `balance` ทางฝั่งขวาของนิพจน์ (ก่อนที่จะมีการกำหนดค่าจริง)
เกิด `UnboundLocalError` ทันที บรรทัด 7–10 ดักข้อผิดพลาดนี้ไว้เพื่อแสดงข้อความอธิบาย และบรรทัด 12 ยืนยันว่าตัวแปร `balance` ระดับ global
ยังคงมีค่า 100 เท่าเดิม เพราะฟังก์ชันไม่สามารถทำงานจนจบได้ ตัวอย่างนี้เป็นข้อผิดพลาดด้าน scope ที่พบบ่อยที่สุดข้อหนึ่งของผู้เริ่มต้นเขียนโปรแกรม

> **หมายเหตุเรื่องเวอร์ชัน Python:** ข้อความ error ข้างต้น (`cannot access local variable ... where it is not associated with a value`)
> เป็นรูปแบบข้อความของ Python 3.11 ขึ้นไป (ยืนยันด้วยการรันจริงบน Python 3.11.15) หากรันด้วย Python เวอร์ชันเก่ากว่า 3.11
> ข้อความจะแตกต่างออกไปเล็กน้อยเป็น `local variable 'balance' referenced before assignment` แต่ประเภทข้อผิดพลาด (`UnboundLocalError`)
> และสาเหตุยังคงเหมือนเดิม

### แหล่งอ้างอิง
- W3Schools — Python Scope (https://www.w3schools.com/python/python_scope.asp): "A variable created inside a function belongs to the local scope of that function" และอธิบายคำสั่ง `global`
- TutorialsPoint — Python Variable Scope (https://www.tutorialspoint.com/python/python_variables_scope.htm): อธิบาย local/global variable และการใช้ `global` เพื่อแก้ไขตัวแปรส่วนกลาง
- GeeksforGeeks — Global and Local Variables in Python (https://www.geeksforgeeks.org/python/global-local-variables-python/): อธิบาย `UnboundLocalError` เมื่อพยายามแก้ไข global โดยไม่ประกาศ `global` ก่อน

---

## แบบฝึกหัดท้ายบท

แบบฝึกหัดชุดนี้มีทั้งหมด **20 ข้อ** ครอบคลุมทั้ง 5 หัวข้อของไฟล์นี้อย่างทั่วถึง (หัวข้อละ 4 ข้อ) โดยผสม 4 ประเภทคำถามในสัดส่วนเท่ากันทุกหัวข้อ
ได้แก่ ข้อเขียนอธิบายแนวคิด ข้อทายผลลัพธ์ (predict output) ข้อหาข้อผิดพลาด (debug) และข้อเขียนโค้ด (coding exercise)
เฉลยทั้งหมดอยู่ในหมวด **"เฉลยแบบฝึกหัด"** ท้ายไฟล์ แยกจากหมวดโจทย์เพื่อไม่ให้เผลอเห็นคำตอบก่อนลองทำเอง
ข้อที่เป็นข้อทายผลลัพธ์และข้อเขียนโค้ดทุกข้อผ่านการรันจริงเพื่อยืนยัน output แล้ว (ดูรายละเอียดในหมวด "การยืนยันผลลัพธ์" ด้านบน)

### 1: จุดประสงค์ของฟังก์ชัน (ข้อ 1–4)

**ข้อ 1 (เขียนอธิบายแนวคิด):** หลักการ DRY (Don't Repeat Yourself) คืออะไร และการใช้ฟังก์ชันช่วยให้ปฏิบัติตามหลักการนี้ได้อย่างไร พร้อมยกตัวอย่างประกอบ

**ข้อ 2 (ทายผลลัพธ์):** จงบอกผลลัพธ์ (output) ที่ได้จากการรันโค้ดต่อไปนี้

```python
01  def multiply(a, b):
02      return a * b
03
04  print(multiply(3, 4))
05  print(multiply(5, 5))
```

**ข้อ 3 (หาข้อผิดพลาด):** โค้ดต่อไปนี้ตั้งใจให้ฟังก์ชันทั้งสองคำนวณค่าดัชนีมวลกาย (BMI) ด้วยสูตรเดียวกัน แต่ผลลัพธ์ที่ได้กลับไม่ตรงกัน จงบอกว่าโค้ดผิดตรงบรรทัดใด และควรแก้ไขอย่างไร

```python
01  def calculate_bmi_v1(weight, height):
02      return weight / (height ** 2)
03
04  def calculate_bmi_v2(weight, height):
05      return weight / height ** 2 * 2
06
07  print(calculate_bmi_v1(70, 1.75))
08  print(calculate_bmi_v2(70, 1.75))
```

**ข้อ 4 (เขียนโค้ด):** จงเขียนฟังก์ชัน `convert_km_to_miles(km)` ที่รับระยะทางเป็นกิโลเมตร แล้วคืนค่าระยะทางเป็นไมล์ (1 กิโลเมตร ≈ 0.621371 ไมล์)
จากนั้นเรียกใช้ฟังก์ชันนี้เพื่อพิมพ์ระยะทาง 10 กิโลเมตร และ 42.195 กิโลเมตร (ระยะมาราธอน) เป็นไมล์

### 2: Function Definition กับ Function Call (ข้อ 5–8)

**ข้อ 5 (เขียนอธิบายแนวคิด):** จงอธิบายความแตกต่างระหว่าง "การนิยามฟังก์ชัน (function definition)" กับ "การเรียกใช้ฟังก์ชัน (function call)"
และอธิบายว่าเพราะเหตุใดการนิยามฟังก์ชันเพียงอย่างเดียวจึงไม่ทำให้คำสั่งภายในฟังก์ชันถูกรัน

**ข้อ 6 (ทายผลลัพธ์):** จงบอกผลลัพธ์ (output) ที่ได้จากการรันโค้ดต่อไปนี้ พร้อมระบุลำดับก่อน-หลัง

```python
01  def show_step(step_number):
02      print(f"ขั้นตอนที่ {step_number}")
03
04  print("เริ่มต้น")
05  show_step(1)
06  show_step(2)
07  print("สิ้นสุด")
```

**ข้อ 7 (หาข้อผิดพลาด):** โค้ดต่อไปนี้รันแล้วเกิดข้อผิดพลาด จงบอกว่าผิดตรงไหนและควรแก้ไขอย่างไร

```python
01  greet_user()
02
03  def greet_user():
04      print("ยินดีต้อนรับ")
```

**ข้อ 8 (เขียนโค้ด):** จงเขียนฟังก์ชัน `print_divider()` ที่พิมพ์เครื่องหมาย `-` จำนวน 20 ตัวติดกัน จากนั้นเรียกใช้ฟังก์ชันนี้คั่นระหว่างข้อความ
`"หัวข้อที่ 1"` และ `"หัวข้อที่ 2"` โดยให้มีเส้นคั่นก่อนข้อความแรก ระหว่างสองข้อความ และหลังข้อความสุดท้ายด้วย (รวมเรียกฟังก์ชัน 3 ครั้ง)

### 3: Parameter กับ Argument (ข้อ 9–12)

**ข้อ 9 (เขียนอธิบายแนวคิด):** พารามิเตอร์ (parameter) กับอาร์กิวเมนต์ (argument) ต่างกันอย่างไร และอาร์กิวเมนต์ตามตำแหน่ง (positional argument)
มีหลักการผูกค่าเข้ากับพารามิเตอร์อย่างไร

**ข้อ 10 (ทายผลลัพธ์):** จงบอกผลลัพธ์ (output) ที่ได้จากการรันโค้ดต่อไปนี้

```python
01  def calculate_power(base, exponent):
02      return base ** exponent
03
04  print(calculate_power(2, 3))
05  print(calculate_power(3, 2))
```

**ข้อ 11 (หาข้อผิดพลาด):** โค้ดต่อไปนี้รันแล้วเกิดข้อผิดพลาด จงบอกว่าผิดตรงไหนและควรแก้ไขอย่างไร

```python
01  def introduce_pet(name, species, age):
02      print(f"{name} เป็น {species} อายุ {age} ปี")
03
04  introduce_pet("Milo", "แมว")
```

**ข้อ 12 (เขียนโค้ด):** จงเขียนฟังก์ชัน `calculate_discounted_price(price, discount_percent)` ที่คืนค่าราคาหลังหักส่วนลด
แล้วเรียกใช้ฟังก์ชันนี้ด้วย**อาร์กิวเมนต์แบบระบุชื่อ (keyword argument)** สำหรับราคา 500 บาท ส่วนลด 10% แล้วพิมพ์ผลลัพธ์

### 4: Return Value กับ Procedure (ข้อ 13–16)

**ข้อ 13 (เขียนอธิบายแนวคิด):** ฟังก์ชันที่มีค่าที่ส่งกลับ (return value) กับฟังก์ชันแบบ procedure (void function) ต่างกันอย่างไร
และหากเก็บผลลัพธ์จากฟังก์ชันแบบ procedure ไว้ในตัวแปร ตัวแปรนั้นจะมีค่าเป็นอะไร

**ข้อ 14 (ทายผลลัพธ์):** จงบอกผลลัพธ์ (output) ที่ได้จากการรันโค้ดต่อไปนี้ ทีละบรรทัด

```python
01  def calculate_average(a, b):
02      return (a + b) / 2
03
04  def display_average(a, b):
05      print((a + b) / 2)
06
07  x = calculate_average(10, 20)
08  y = display_average(10, 20)
09  print(x)
10  print(y)
```

**ข้อ 15 (หาข้อผิดพลาด):** โค้ดต่อไปนี้รันแล้วเกิดข้อผิดพลาด จงบอกว่าผิดตรงไหนและควรแก้ไขอย่างไร

```python
01  def get_max_score(scores):
02      print(max(scores))
03
04  result = get_max_score([70, 85, 90])
05  total = result + 10
06  print(total)
```

**ข้อ 16 (เขียนโค้ด):** จงเขียนฟังก์ชัน `count_vowels(text)` ที่รับสตริงข้อความภาษาอังกฤษ แล้ว**คืนค่า**จำนวนสระ (`a`, `e`, `i`, `o`, `u` ตัวพิมพ์เล็ก)
ที่พบในข้อความนั้น (ฟังก์ชันต้องมี return ไม่ใช่แค่ print) จากนั้นเรียกใช้กับคำว่า `"education"` แล้วพิมพ์ผลลัพธ์

### 5: ขอบเขตตัวแปร (Scope) (ข้อ 17–20)

**ข้อ 17 (เขียนอธิบายแนวคิด):** เพราะเหตุใดการ**แก้ไข**ค่าตัวแปรส่วนกลาง (global variable) จากภายในฟังก์ชันจึงต้องประกาศคำสั่ง `global` ก่อนเสมอ
ในขณะที่การ**อ่าน**ค่าตัวแปรส่วนกลางไม่จำเป็นต้องประกาศอะไรเพิ่มเติม

**ข้อ 18 (ทายผลลัพธ์):** จงบอกผลลัพธ์ (output) ที่ได้จากการรันโค้ดต่อไปนี้

```python
01  x = 10
02
03  def show_x():
04      x = 99
05      print(x)
06
07  show_x()
08  print(x)
```

**ข้อ 19 (หาข้อผิดพลาด):** โค้ดต่อไปนี้รันแล้วเกิดข้อผิดพลาด จงบอกว่าผิดตรงไหนและควรแก้ไขอย่างไร

```python
01  counter = 0
02
03  def increment():
04      counter = counter + 1
05      print(counter)
06
07  increment()
```

**ข้อ 20 (เขียนโค้ด):** จงเขียนฟังก์ชัน `add_point(points)` ที่นำค่า `points` ไปบวกเพิ่มในตัวแปรส่วนกลางชื่อ `score` (เริ่มต้นที่ 0)
โดยใช้คำสั่ง `global` ให้ถูกต้อง จากนั้นเรียกใช้ฟังก์ชันนี้เพื่อบวกคะแนน 15 และ 20 ตามลำดับ แล้วพิมพ์ค่า `score` สุดท้าย

---

> **หมายเหตุ:** เฉลยของแบบฝึกหัดในสมุดบันทึกนี้ถูกย้ายไปไว้ในไฟล์ `lecture_note_basic_answers.ipynb` (อยู่ในโฟลเดอร์เดียวกัน) เปิดไฟล์นั้นเมื่อทำแบบฝึกหัดเสร็จแล้วและต้องการตรวจคำตอบ